src/rust/bitbox02-rust/src/hww/api/show_mnemonic.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate alloc;
+use alloc::vec::Vec;
+
+use super::pb;
+use super::Error;
+
+use pb::response::Response;
+
+use crate::workflow::{mnemonic, status, unlock};
+use bitbox02::keystore;
+
+const NUM_RANDOM_WORDS: u8 = 5;
+
+fn create_random_unique_words(word: &str, length: u8) -> (u8, Vec<&str>) {
+    fn rand16() -> u16 {
+        let mut rand = [0u8; 32];
+        bitbox02::random::mcu_32_bytes(&mut rand);
+        ((rand[0] as u16) << 8) | (rand[1] as u16)
+    }
+
+    let index_word = (rand16() as u8) % length;
+    let mut picked_indices = Vec::new();
+    let result = (0..length)
+        .map(|i| {
+            // The correct word at the right index.
+            if i == index_word {
+                return word;
+            }
+
+            // A random word everywhere else.
+            // Loop until we get a unique word, we don't want repeated words in the list.
+            loop {
+                let idx = rand16() % keystore::BIP39_WORDLIST_LEN;
+                if picked_indices.contains(&idx) {
+                    continue;
+                };
+                let random_word = keystore::get_bip39_word(idx).unwrap();
+                if random_word == word {
+                    continue;
+                }
+                picked_indices.push(idx);
+                return random_word;
+            }
+        })
+        .collect();
+
+    (index_word, result)
+}
+
+pub async fn process() -> Result<Response, Error> {
+    unlock::unlock_keystore("Unlock device").await?;
+
+    let mnemonic_sentence = keystore::get_bip39_mnemonic()?;
+
+    let words: Vec<&str> = mnemonic_sentence.split(' ').collect();
+
+    // Part 1) Scroll through words
+    mnemonic::show_mnemonic(&words).await?;
+
+    // Part 2) Confirm words
+    for (word_idx, word) in words.iter().enumerate() {
+        let title = format!("{:02}", word_idx + 1);
+        let (correct_idx, mut choices) = create_random_unique_words(word, NUM_RANDOM_WORDS);
```

Comment:
Nothing against vectors, but another option is to pass a `&mut [Option<&str>, 5]` to avoid allocations.

Reply:
aha, I didn't consider FFI so you can keep it. BUT now I will shock you with my rust-foo. `Option` has null pointer optimizations, and since &str cannot be null, null is used to represent `None`.

So it would be equivalent to send `Option<&str>` to FFI as `&str`.

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1321987a22c83a654259e17e6fab73b7

If None needed it's own storage it would be larger than 16.

Reply:
okay that is indeed amazing rust-foo :D is that guaranteed or an implementation detail?

I forgot that it is not sent over FFI directly as ptr, but all strings are str_to_cstr'd (copied into a new buffer with a null terminator). 

But anyway, the menu_create() takes `&[&str]`. Are you saying I can coerce `&[Option<&str>; 5]` to `&[&str]` safely if I know that all the entries are `Some()`? :scream:

Reply:
It's practically guaranteed, otherwise `Option<>` would add a lot of overhead to types that don't need it.

No I'm not saying you should do that, you need to unwrap the options if you stay in rust.

But wait a minute, why don't you just do this? `fn create_random_unique_words() -> (u8, [&'static str; NUM_RANDOM_WORDS])`

Reply:
At the result site I am pushing one more element to the result.

Reply:
`fn create_random_unique_words() -> (u8, [&'static str; NUM_RANDOM_WORDS+1])`?

Reply:
breaks modularity? :joy: 

but no I get it, it is definitely possible to do it without allocations. But imho Vec<> is so ergonomic and I see no downside in using it.

Reply:
👍 allocations aren't so bad when you have "destructors" :)